OAuth 2.0 Device Authorization Grant (#4341)
matrix-org/matrix-spec-proposals: proposals/4341-device-authorization-grant.md

# MSC4341: Support for RFC 8628 Device Authorization Grant

The current [OAuth 2.0 API](https://spec.matrix.org/v1.15/client-server-api/#oauth-20-api) requires the user to complete
authentication using a web browser on the device where the Matrix client is running.

This can be problematic if the device does not have a built in web browser or the user wishes to use a different device
to complete login.

> [RFC 8628](https://datatracker.ietf.org/doc/html/rfc8628) defines the OAuth 2.0 Device Authorization Grant which can be

Author: Most of the following links are to specific questions which don't appear to be supported by reusable (aka reference) links

used for this purpose.

## Proposal

Add the [RFC 8628 OAuth 2.0 Device Authorization Grant](https://datatracker.ietf.org/doc/html/rfc8628) to the [list of supported
grant types](https://spec.matrix.org/v1.15/client-server-api/#grant-types) in the Client-Server API.

This grant requires the client to know the following authorization server metadata:

- `grant_types_supported` - this would include `urn:ietf:params:oauth:grant-type:device_code`
- `device_authorization_endpoint` - this would be added to the [table of fields for the 200 response](https://spec.matrix.org/v1.15/client-server-api/#server-metadata-discovery)
- `token_endpoint` - this is already included in the spec

> To use this grant, homeservers and clients MUST:

Author: I was thinking that it is optional.

But, in turn I think I would make the authorization code grant optional too.

Then the Matrix spec would say that the homeserver must provide at least one of: authorization code grant; device authorization grant;

The refresh token grant would remain mandatory as it is very sensible from security point of view.

> To use this grant, homeservers and clients MUST:

Author: I've rolled these in to https://github.com/matrix-org/matrix-spec-proposals/commit/68184aac3110b985b9457328d916e8a7df2a35d2

# MSC4341: Support for RFC 8628 Device Authorization Grant

The current [OAuth 2.0 API](https://spec.matrix.org/v1.15/client-server-api/#oauth-20-api) requires the user to complete
authentication using a web browser on the device where the Matrix client is running.

This can be problematic if the device does not have a built in web browser or the user wishes to use a different device.

> It would also be useful in scenarios where catching the redirect back to the client is hard, like in CLI apps, or
> desktop apps with no redirect custom schemes.

Reviewer: It's slightly unclear to me what "it" refers to here. I suspect you mean the device authorization grant but that is only introduced in the next paragraph. Maybe rephrasing it as a problem rather than something helpful would be clearer?

```suggestion
It also causes issues in scenarios where catching the redirect back to the client is hard, like in CLI apps, or
desktop apps with no redirect custom schemes.
```

used for this purpose.

## Proposal

Add the [RFC 8628 OAuth 2.0 Device Authorization Grant](https://datatracker.ietf.org/doc/html/rfc8628) to the [list of supported
grant types](https://spec.matrix.org/v1.15/client-server-api/#grant-types) in the Client-Server API.

This grant requires the client to know the following authorization server metadata:

- `grant_types_supported` - this would include `urn:ietf:params:oauth:grant-type:device_code`
- `device_authorization_endpoint` - this would be added to the [table of fields for the 200 response](https://spec.matrix.org/v1.15/client-server-api/#server-metadata-discovery)
- `token_endpoint` - this is already included in the spec

To use this grant, homeservers and clients MUST:

- Support the device authorization grant as per RFC 8628.
- Support the refresh token grant.

The [login flow section](https://spec.matrix.org/v1.15/client-server-api/#login-flow) would need to be updated to
reflect that more than one login grant type exists.

As with the existing authorization code grant, when authorization is granted to a client, the homeserver MUST issue a
refresh token to the client in addition to the access token.

> The access token MUST be short-lived and SHOULD be refreshed using the `refresh_token` when expired.

Reviewer: The current spec also includes this phrase but what exactly is "short-lived"? The RFC uses the same language for authorization codes but I think you'd typically want access token lifetimes to have a significantly larger lifetime than auth codes?

For example:

```http
HTTP/1.1 200 OK
Content-Type: application/json

{
  "device_code": "GmRhmhcxhwAzkoEqiMEg_DnyEysNkuNhszIySk9eS",
  "user_code": "WDJB-MJHT",
  "verification_uri": "https://account.matrix.org/link",
  "verification_uri_complete": "https://account.matrix.org/link?user_code=WDJB-MJHT",
  "expires_in": 1800,
  "interval": 5
}
```

It is recommended that the server provides a `verification_uri_complete` such that the user does not need to type in the
`user_code`.

5. The Matrix client device conveys the returned `verification_uri_complete` (and/or `verification_uri`+`user_code`) to
the user.

Exactly how the client does this depends on the specific device characteristics and use case.

For example, for a CLI application the `verification_uri` and `user_code` could be displayed as text for the user to
type into their other device.

Or, another example would be the `verification_uri_complete` (with fallback to the `verification_uri`) could be encoded
and displayed as a QR code for scanning on the other device (e.g. on a TV).

> 6. The user visits opens the `verification_uri_complete` (or `verification_uri`) on their other device that has a web browser.

Reviewer: ```suggestion
6. The user opens the `verification_uri_complete` (or `verification_uri`) on their other device that has a web browser.
```
